text/0000-macros-1.1.md

```diff
@@ -0,0 +1,552 @@
+
+#[rustc_macro_derive(Serialize)]
+pub fn derive_serialize(_cx: &mut Context,
+                        input: TokenStream) -> TokenStream {
+    let input = input.to_source();
+
+    // use syntex_syntax from crates.io to parse `input` into an AST
+
+    // use this AST to generate an impl of the `Serialize` trait for the type in
+    // question
+
+    // convert that impl to a string
+
+    // parse back into a token stream
+    return TokenStream::from_source(&impl_source).unwrap()
+}
+```
+
+Next, crates will depend on this such as:
+
+```toml
+[dependencies]
+serde = "0.9"
+serde-macros = "0.9"
+```
+
+And finally use it as such:
+
+```rust
+extern crate serde;
+extern crate serde_macros;
+
+#[derive(Serialize)]
+pub struct Foo {
+    a: usize,
+    #[serde(rename = "foo")]
+    b: String,
+}
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This is not an interface that would be considered for stabilization in a void,
+  there are a number of known drawbacks to the current macro system in terms of
+  how it architecturally fits into the compiler. Additionally, there's work
+  underway to solve all these problems with macros 2.0.
+
+  As mentioned before, however, the stable version of macros 2.0 is currently
+  quite far off, and the desire for features like custom derive are very real
+  today. The rationale behind this RFC is that the downsides are an acceptable
+  tradeoff from moving a significant portion of the nightly ecosystem onto stable
+  Rust.
+
+* This implementation is likely to be less performant than procedural macros
+  are today. Round tripping through strings isn't always a speedy operation,
+  especially for larger expansions. Strings, however, are a very small
+  implementation detail that's easy to see stabilized until the end of time.
+  Additionally, it's planned to extend the `TokenStream` API in the future to
+  allow more fine-grained transformations without having to round trip through
+  strings.
+
+* Users will still have an inferior experience to today's nightly macros
+  specifically with respect to compile times. The `syntex_syntax` crate takes
+  quite a few seconds to compile, and this would be required by any crate which
+  uses serde. To offset this, though, the `syntex_syntax` could be *massively*
+  stripped down as all it needs to do is parse struct declarations mostly. There
+  are likely many other various optimizations to compile time that can be
+  applied to ensure that it compiles quickly.
+
+* Plugin authors will need to be quite careful about the code which they
+  generate as working with strings loses much of the expressiveness of macros in
+  Rust today. For example:
+
+  ```rust
+  macro_rules! foo {
+      ($x:expr) => {
+          #[derive(Serialize)]
+          enum Foo { Bar = $x, Baz = $x * 2 }
+      }
+  }
+  foo!(1 + 1);
+  ```
+
+  Plugin authors would have to ensure that this is not naively interpreted as
+  `Baz = 1 + 1 * 2` as this will cause incorrect results.
```

Comment:
Couldn't we add parenthesis around all such fragments at stringification time to avoid the misinterpretation?
Plugin authors can't do anything about this if all they have is conversion to `String`.